e3sm_diags/driver/utils/general.py

```diff
@@ -15,3 +15,68 @@ def _monotonically_decreasing(L):
 def _monotonically_increasing(L):
     # FIXME: B905: zip() without an explicit strict= parameter
     return all(x <= y for x, y in zip(L, L[1:]))
+
+
+def pad_year(year: int | str) -> str:
+    """Pad the year with leading zeros to ensure it is 4 digits.
+
+    This function ensures that the input year is properly formatted as a
+    4-digit string, which is required for ISO-8601 date formats (YYYY-MM-DD).
+    If the year is less than 1000, it is padded with leading zeros.
+
+    Parameters
+    ----------
+    year : int or str
+        The year to pad. Must be a non-negative integer or a string representing
+        a non-negative integer. Floats are not allowed.
+
+    Returns
+    -------
+    str
+        The padded year as a 4-digit string (e.g., "0042" for year 42).
+
+    Raises
+    ------
+    ValueError
+        If the input year is not a non-negative integer or a string representing
+        a non-negative integer, or if it is a float, or if it is outside the
+        range 0 to 9999 inclusive.
+
+    Examples
+    --------
+    >>> pad_year(42)
+    '0042'
+    >>> pad_year("42")
+    '0042'
+    >>> pad_year(2023)
+    '2023'
+    >>> pad_year("2023")
+    '2023'
+    >>> pad_year(-1)
+    Traceback (most recent call last):
+        ...
+    ValueError: Year must be between 0 and 9999 inclusive.
+    >>> pad_year(10000)
+    Traceback (most recent call last):
+        ...
+    ValueError: Year must be between 0 and 9999 inclusive.
+    >>> pad_year(42.0)
+    Traceback (most recent call last):
+        ...
+    ValueError: Year must not be a float.
+    """
+    try:
+        if isinstance(year, float):
+            raise ValueError("Year must not be a float.")
+
+        year = int(year)
+    except (ValueError, TypeError) as e:
+        raise ValueError(
+            "Year must be a non-negative integer or a string representing a "
+            "non-negative integer."
+        ) from e
+
+    if year < 0 or year > 9999:
+        raise ValueError("Year must be between 0 and 9999 inclusive.")
+
+    return f"{year:04d}"
```

Comment:
The new `pad_year` function to ensure time slicing with Xarray works, even with years < 1000.

Reply:
One thing to note is that we should call `pad_year` whenever we do any subsetting with Xarray's `.sel` on time if start year/end year is used. This ensures that `.sel` works even with start year/end year < 1000.

For now, this solution is good enough. However, we might consider another fix in `CoreParameter` that pads any year attribute automatically to 4 digits because it might be easy to forget to call `pad_year`.